x-pack/plugins/observability_solution/apm/server/routes/service_map/fetch_service_paths_from_trace_ids.ts

```diff
@@ -73,48 +97,46 @@ export async function fetchServicePathsFromTraceIds({
       scripted_metric: {
         params: {
           limit: numDocsPerShardAllowed,
+          fieldsToCopy: SCRIPTED_METRICS_FIELDS_TO_COPY,
         },
         init_script: {
           lang: 'painless',
-          source: `state.docCount = 0;
-          state.limit = params.limit;
-          state.eventsById = new HashMap();
-          String[] fieldsToCopy = new String[] {
-              'parent.id',
-              'service.name',
-              'service.environment',
-              'span.destination.service.resource',
-              'trace.id',
-              'processor.event',
-              'span.type',
-              'span.subtype',
-              'agent.name'
-            };
-            state.fieldsToCopy = fieldsToCopy;`,
+          source: `
+            state.docCount = 0;
+            state.limit = params.limit;
+            state.eventsById = new HashMap();
+            state.fieldsToCopy = params.fieldsToCopy;`,
         },
         map_script: {
           lang: 'painless',
-          source: `if (state.docCount >= state.limit) {
-            return; // Stop processing if the document limit is reached
-          }
-            def id;
-            id = $('span.id', null);
+          source: `
+            if (state.docCount >= state.limit) {
+              /* Stop processing if the document limit is reached */
+              return; 
+            }
+
+            def id = $('span.id', null);
             if (id == null) {
               id = $('transaction.id', null);
             }
 
-            def copy = new HashMap();
-            copy.id = id;
+            /* makes sure we don't process the same event twice */
+            if (id != null && !state.eventsById.containsKey(id)) {
+              def copy = new HashMap();
+              copy.id = id;
 
-            for(key in state.fieldsToCopy) {
-              def value = $(key, null);
-              if (value != null) {
-                copy[key] = value;
+              for(key in state.fieldsToCopy) {
+                def value = $(key, null);
+                if (value != null) {
+                  copy[key] = value;
+                }
               }
+
+              state.eventsById[id] = copy;
             }
 
-            state.eventsById[id] = copy;
-            state.docCount+=1`,
+            state.docCount++;
```

Comment:
@crespocarlos should this be in the check for duplicates? as in only incremented if its unique and added to the state?  otherwise we might reach our limit while the state actually doesn't have many docs to return if there are many duplicates.

Reply:
yeah, let's move it into the duplicates check `if`. It makes sense to increment when unique docs are found instead of when `map_script` is called.